opennms-webapp-rest/src/main/java/org/opennms/web/rest/v2/ApplicationStatusRestService.java

```diff
@@ -0,0 +1,281 @@
+                ++i;
+            }
+            --i;
+
+            resultList.add(mergedInterval);
+        }
+
+        return resultList;
+    }
+
+    private boolean intervalsOverlap(final DowntimeInterval donwtimeInterval1, final DowntimeInterval donwtimeInterval2) {
+        return !(Math.max(donwtimeInterval1.start, donwtimeInterval2.start) > Math.min(donwtimeInterval1.end, donwtimeInterval2.end));
+    }
+
+    private double calculateServicePercentageUptime(final Collection<OnmsLocationSpecificStatus> statusChanges, final List<DowntimeInterval> downtimeIntervals, final long start, final long end) {
+        double totalTimeMillis = end - start;
+        double uptimeMillis = totalTimeMillis;
+
+        OnmsLocationSpecificStatus lastChange = null;
+
+        for(final OnmsLocationSpecificStatus status : statusChanges) {
+            long s = (lastChange == null ? 0 : lastChange.getPollResult().getTimestamp().getTime());
+            long e = status.getPollResult().getTimestamp().getTime();
+            if (lastChange != null && lastChange.getPollResult().isDown() ) {
+                if(s < e) {
+                    uptimeMillis -= (Math.min(e, end)-Math.max(s, start));
+                    downtimeIntervals.add(new DowntimeInterval(Math.max(s, start), Math.min(e, end)));
+                }
+            }
+            lastChange = status;
+        }
+
+        if (lastChange != null && lastChange.getPollResult().isDown() ) {
+            long s = lastChange.getPollResult().getTimestamp().getTime();
+            long e = end;
+            if(s < e) {
+                uptimeMillis -= (Math.min(e, end)-Math.max(s, start));
+                downtimeIntervals.add(new DowntimeInterval(Math.max(s, start), Math.min(e, end)));
+            }
+        }
+
+        return uptimeMillis / totalTimeMillis;
+    }
+
+    public double calculateApplicationPercentageUptime(final List<DowntimeInterval> intervals, final long start, final long end) {
+        double totalTimeMillis = end - start;
+        double uptimeMillis = totalTimeMillis;
+        for(final DowntimeInterval interval : intervals) {
+            uptimeMillis -= (Math.min(interval.end, end)-Math.max(interval.start, start));
+        }
+        return uptimeMillis / totalTimeMillis;
+    }
+
+
+    private ApplicationStatus getApplicationStatus(final OnmsApplication onmsApplication, final Collection<OnmsLocationSpecificStatus> statusChanges, final long start, final long end) {
+        final Map<OnmsMonitoringLocation, Map<OnmsMonitoredService, List<OnmsLocationSpecificStatus>>> m = new HashMap<>();
+
+        for(final OnmsLocationSpecificStatus onmsLocationSpecificStatus : statusChanges) {
+            m.computeIfAbsent(onmsLocationSpecificStatus.getLocation(), k -> new HashMap<>())
+                    .computeIfAbsent(onmsLocationSpecificStatus.getMonitoredService(), k -> new ArrayList<>())
+                    .add(onmsLocationSpecificStatus);
+        }
```

Comment:
Is this collection required? Why not just calculate the values and fill the `ApplicationStatus`?